src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/RasterDownload.java

```diff
@@ -178,170 +185,441 @@ public Resource execute(String mimeType, final ProgressListener progressListener
-                                + "This raster contains "+sampleDimensionsNumber+" band"
-                                + (sampleDimensionsNumber>1?"s":""));            
-                    }
+            if (targetSizeX == null && targetSizeY == null) {
+                // No size is specified. Just do a read and reproject (if needed) + a final crop
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    LOGGER.log(Level.FINE, "No Target size has been specified. Requested GridGeometry "
+                            + "will be automatically computed");
                 }
-                BandSelectProcess bandSelectProcess = new BandSelectProcess();
+                GridGeometryProvider provider = new GridGeometryProvider(reader, roiManager, filter);
+                requestedGridGeometry = provider.getGridGeometry();
                 
-                //using null for the VisibleSampleDimension parameter of BandSelectProcess.execute. 
-                //GeoTools BandSelector2D takes care of remapping visible band index
-                //or assigns it to first band in order if remapping is not possible
-                bandFilteredCoverage = bandSelectProcess.execute(
-                        originalGridCoverage, bandIndices, null);
-            
-            }else{
-                bandFilteredCoverage = originalGridCoverage;
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    LOGGER.log(Level.FINE, "Computed requested GridGeometry: " + requestedGridGeometry.toString());
+                }
+                readParameters = CoverageUtils.mergeParameter(parameterDescriptors, readParameters, 
+                        requestedGridGeometry, AbstractGridFormat.READ_GRIDGEOMETRY2D.getName().getCode());
+            } else {
+                if (targetSizeX == null || targetSizeY == null) {
+                    // one of the 2 sizes is not specified. Delegate
+                    // scaleToTarget to compute the second one.
+                    ScaleToTarget scaling = new ScaleToTarget(reader);
+                    scaling.setTargetSize(targetSizeX, targetSizeY);
+                    Integer[] computedSizes = scaling.getTargetSize();
+                    targetSizeX = computedSizes[0];
+                    targetSizeY = computedSizes[1];
+                }
+
+                // Since we have imposed a target size, delegate GridCoverageRenderer to do all the dirty job
+                requestedGridGeometry = new GridGeometry2D(
+                        new GridEnvelope2D(0, 0, targetSizeX, targetSizeY), targetEnvelope);
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    LOGGER.log(Level.FINE, "Target size has been specified. Setting up requested GridGeometry: "
+                                    + requestedGridGeometry.toString());
+                }
             }
 
+            readParameters = updateReadParams(readParameters, parameterDescriptors, bandIndices, filter);
```

Comment:
A comment explaining "update how" would be nice (e.g., add filter, enforce deferred loading and the like)